common/persistence/visibility/factory.go

```diff
@@ -0,0 +1,244 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package visibility
+
+import (
+	"go.temporal.io/server/common/config"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/persistence/visibility/manager"
+	"go.temporal.io/server/common/persistence/visibility/store"
+	"go.temporal.io/server/common/persistence/visibility/store/cassandra"
+	"go.temporal.io/server/common/persistence/visibility/store/elasticsearch"
+	esclient "go.temporal.io/server/common/persistence/visibility/store/elasticsearch/client"
+	"go.temporal.io/server/common/persistence/visibility/store/sql"
+	"go.temporal.io/server/common/resolver"
+	"go.temporal.io/server/common/searchattribute"
+)
+
+func NewManager(
+	persistenceCfg config.Persistence,
+	persistenceResolver resolver.ServiceResolver,
+
+	defaultIndexName string,
+	esClient esclient.Client,
+	esProcessorConfig *elasticsearch.ProcessorConfig,
+	searchAttributesProvider searchattribute.Provider,
+	searchAttributesMapper searchattribute.Mapper,
+
+	standardVisibilityPersistenceMaxReadQPS dynamicconfig.IntPropertyFn,
+	standardVisibilityPersistenceMaxWriteQPS dynamicconfig.IntPropertyFn,
+	advancedVisibilityPersistenceMaxReadQPS dynamicconfig.IntPropertyFn,
+	advancedVisibilityPersistenceMaxWriteQPS dynamicconfig.IntPropertyFn,
+	enableAdvancedVisibilityRead dynamicconfig.BoolPropertyFnWithNamespaceFilter,
+	advancedVisibilityWritingMode dynamicconfig.StringPropertyFn,
```

Comment:
consider wrap this within a struct? e.g. VisibilityOptions

Reply:
I was thinking about it but I also has an idea of splitting this manager into two: read and write. History service does only writes and frontend only reads. And then these 6 parameters will also split into 3+3.
Btw, should I just inject dynamic config client instead and read whatever I need from the manger itself?